### ISSUE TITLE
Nightly job failures, Hip and Cuda with gcc/8.5.0 in the toolchain: linking issues related to std::filesystem

### ISSUE THREAD

**Issue by ndellingwood** [kokkos/kokkos#7416]
**Describe the bug**

Nightly Hip jobs, with gcc/8.5.0 on the system path, have linkage issue related to std::filesystem:
```
20:59:25 ld.lld: error: undefined symbol: std::filesystem::temp_directory_path[abi:cxx11]()
20:59:25 >>> referenced by TestHIP_InterOp_Graph.cpp
20:59:25 >>>               CMakeFiles/Kokkos_CoreUnitTest_HIPInterOpGraph.dir/hip/TestHIP_InterOp_Graph.cpp.o:((anonymous namespace)::hip_graph_instantiate_and_debug_dot_print_Test::TestBody())
20:59:25 
20:59:25 ld.lld: error: undefined symbol: std::filesystem::status(std::filesystem::__cxx11::path const&)
20:59:25 >>> referenced by TestHIP_InterOp_Graph.cpp
20:59:25 >>>               CMakeFiles/Kokkos_CoreUnitTest_HIPInterOpGraph.dir/hip/TestHIP_InterOp_Graph.cpp.o:((anonymous namespace)::hip_graph_instantiate_and_debug_dot_print_Test::TestBody())
20:59:25 
20:59:25 ld.lld: error: undefined symbol: std::filesystem::file_size(std::filesystem::__cxx11::path const&)
20:59:25 >>> referenced by TestHIP_InterOp_Graph.cpp
20:59:25 >>>               CMakeFiles/Kokkos_CoreUnitTest_HIPInterOpGraph.dir/hip/TestHIP_InterOp_Graph.cpp.o:((anonymous namespace)::hip_graph_instantiate_and_debug_dot_print_Test::TestBody())
20:59:25 
20:59:25 ld.lld: error: undefined symbol: std::filesystem::__cxx11::path::_M_split_cmpts()
20:59:25 >>> referenced by TestHIP_InterOp_Graph.cpp
20:59:25 >>>               CMakeFiles/Kokkos_CoreUnitTest_HIPInterOpGraph.dir/hip/TestHIP_InterOp_Graph.cpp.o:(std::filesystem::__cxx11::path::path<char [14], std::filesystem::__cxx11::path>(char const (&) [14], std::filesystem::__cxx11::path::format))
20:59:25 >>> referenced by TestHIP_InterOp_Graph.cpp
20:59:25 >>>               CMakeFiles/Kokkos_CoreUnitTest_HIPInterOpGraph.dir/hip/TestHIP_InterOp_Graph.cpp.o:(std::filesystem::__cxx11::path::operator/=(std::filesystem::__cxx11::path const&))
20:59:25 
20:59:25 ld.lld: error: undefined symbol: std::filesystem::__cxx11::path::has_root_directory() const
20:59:25 >>> referenced by TestHIP_InterOp_Graph.cpp
20:59:25 >>>               CMakeFiles/Kokkos_CoreUnitTest_HIPInterOpGraph.dir/hip/TestHIP_InterOp_Graph.cpp.o:(std::filesystem::__cxx11::path::operator/=(std::filesystem::__cxx11::path const&))
20:59:25 
20:59:25 ld.lld: error: undefined symbol: std::filesystem::__cxx11::path::has_filename() const
20:59:25 >>> referenced by TestHIP_InterOp_Graph.cpp
20:59:25 >>>               CMakeFiles/Kokkos_CoreUnitTest_HIPInterOpGraph.dir/hip/TestHIP_InterOp_Graph.cpp.o:(std::filesystem::__cxx11::path::operator/=(std::filesystem::__cxx11::path const&))
20:59:25 clang: error: linker command failed with exit code 1 (use -v to see invocation)
20:59:25 make[2]: *** [core/unit_test/CMakeFiles/Kokkos_CoreUnitTest_HIPInterOpGraph.dir/build.make:122: core/unit_test/Kokkos_CoreUnitTest_HIPInterOpGraph] Error 1
```

A couple more notes:

* This issue has not occurred with all jobs using gcc/8, I'm not sure yet why the discrepancy
* Might be related to this note from [cppreference.com](https://en.cppreference.com/w/cpp/filesystem)

> Using this library may require additional compiler/linker options. GNU implementation prior to 9.1 requires linking with -lstdc++fs and LLVM implementation prior to LLVM 9.0 requires linking with -lc++fs. 

This began occurring following merge of these commits:

**Changes**

    Bump actions/cache from 4.0.2 to 4.1.0 ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_carawaymi210_rocm600_hip_serial/187/changes#efdd419d4588f870edf89f13af2cddb8e145c953) / [githubweb](https://github.com/kokkos/kokkos/commit/efdd419d4588f870edf89f13af2cddb8e145c953))
    Bump github/codeql-action from 3.26.9 to 3.26.11 ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_carawaymi210_rocm600_hip_serial/187/changes#7ec89289efdebb3332638d0efbb74190151e98e0) / [githubweb](https://github.com/kokkos/kokkos/commit/7ec89289efdebb3332638d0efbb74190151e98e0))
    graph: allow access to native graph object ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_carawaymi210_rocm600_hip_serial/187/changes#54980770f44beb8799f5deb47fdfb0b827912425) / [githubweb](https://github.com/kokkos/kokkos/commit/54980770f44beb8799f5deb47fdfb0b827912425))
    Use hipMallocAsync when using HIP execution space (#7324) ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_carawaymi210_rocm600_hip_serial/187/changes#d188dd5440bc883245bb6e4e7512e8c2f3148e3d) / [githubweb](https://github.com/kokkos/kokkos/commit/d188dd5440bc883245bb6e4e7512e8c2f3148e3d))
    Refactor: Remove unnecessary `Mirror*Types` structs (#7339) ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_carawaymi210_rocm600_hip_serial/187/changes#fdb75ebcc7aa61e7d67379cd161605350ed6f7d4) / [githubweb](https://github.com/kokkos/kokkos/commit/fdb75ebcc7aa61e7d67379cd161605350ed6f7d4))

**Please include the following for a minimal reproducer**

**Reproducer notes** (caraway MI210 queue):
```bash
salloc -N 1 -p MI210

module load cmake rocm/6.0.0

${KOKKOS_PATH}/generate_makefile.bash --with-hip --with-serial --compiler=hipcc --arch=VEGA90A --kokkos-path=${KOKKOS_PATH}

make -j16
```

**Comment by dalg24** [comment]
This comes from https://github.com/kokkos/kokkos/pull/6904

**Comment by romintomasetti** [comment]
I couldn't reproduce this on `Godbolt`, but here is a `Docker` reproducer:
```dockerfile
# Try me with
#   docker buildx build --tag test - < dockerfile.repro
FROM gcc:8.5.0

COPY <<EOF test.cpp
#include <filesystem>
#include <iostream>

int main()
{
    const std::filesystem::path filename("hello.dot");
    const auto my_path = std::filesystem::temp_directory_path() / filename;
    std::cout << my_path << std::endl;
}
EOF

RUN g++ -std=c++17 test.cpp -o test.exe && ./test.exe
```
It crashes with `test.cpp:(.text+0x2d): undefined reference to `std::filesystem::temp_directory_path[abi:cxx11]()'` (and other related stuff).

**Comment by dalg24** [comment]
>  I couldn't reproduce this on `Godbolt`

I just copy and pasted your code and I see the link time error with GCC 8.5 and it is =fine with 9.X
https://godbolt.org/z/r3vE58M3a

**Comment by romintomasetti** [comment]
Here are the results of my investigation (`0` means it worked, see code below):
```bash
INFO:root:GCC 8.5.0 returns 1.
INFO:root:GCC 9.0.0 returns 1.
INFO:root:GCC 9.1.0 returns 0.
INFO:root:GCC 9.2.0 returns 0.
INFO:root:GCC 12.3.0 returns 0.
INFO:root:GCC 13.2.0 returns 0.
INFO:root:LLVM 8 returns 1.
INFO:root:LLVM 9 returns 1.
INFO:root:LLVM 10 returns 1.
INFO:root:LLVM 11 returns 0.
INFO:root:LLVM 12 returns 0.
```

So it seems that for `LLVM`, it is not *prior* to `9`, but rather `9` included.

<details>

```dockerfile
ARG BASE_IMAGE

FROM ${BASE_IMAGE}

COPY <<EOF test.cpp
#include <filesystem>
#include <iostream>

int main()
{
    const std::filesystem::path filename("hello.dot");
    const auto my_path = std::filesystem::temp_directory_path() / filename;
    std::cout << my_path << std::endl;
}
EOF

ARG COMPILER

RUN ${COMPILER} --version && ${COMPILER} -std=c++17 test.cpp -o test.exe && ./test.exe
```

```python
import logging
import subprocess

def build(base_image : str, compiler : str, dockerfile : str):
    return subprocess.run([
        'docker', 'buildx', 'build', '--tag=test',
        '--build-arg', 'BASE_IMAGE=' + base_image,
        '--build-arg', 'COMPILER=' + compiler,
        '--file', dockerfile,
        '.',
    ], stdout = subprocess.PIPE, stderr = subprocess.STDOUT)

def check_gcc(dockerfile : str):
    """Check which GCC versions are to be skipped."""
    for version in ['8.5.0', '9.0.0', '9.1.0', '9.2.0', '12.3.0', '13.2.0']:
        result = build(base_image = f'gcc:{version}', compiler = 'g++', dockerfile = dockerfile)
        logging.info(f"GCC {version} returns {result.returncode}.")

def check_llvm(dockerfile : str):
    """Check which LLVM versions are to be skipped."""
    for version in ['8', '9', '10', '11', '12']:
        result = build(base_image = f'silkeh/clang:{version}', compiler = 'clang++', dockerfile = dockerfile)
        logging.info(f"LLVM {version} returns {result.returncode}.")

if __name__ == '__main__':

    logging.basicConfig(level = logging.INFO)

    dockerfile = 'dockerfile.repro'

    check_gcc(dockerfile = dockerfile)

    check_llvm(dockerfile = dockerfile)
```

</details>

**Comment by dalg24** [comment]
(Guards were wrong and it is likely associated to the standard library version not the compiler)

**Comment by dalg24** [comment]
Fixed by #7421
